src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1057 @@
+ */
+static inline int vaplic_vcpuid_to_pcpuid(struct vcpu *vcpu){
+    return vm_translate_to_pcpuid(vcpu->vm, vcpu->id);
+}
+
+static uint32_t vaplic_get_domaincfg(struct vcpu *vcpu);
+static uint32_t vaplic_get_target(struct vcpu *vcpu, irqid_t intp_id); 
+static uint32_t vaplic_get_idelivery(struct vcpu *vcpu, uint16_t idc_id);
+static uint32_t vaplic_get_iforce(struct vcpu *vcpu, uint16_t idc_id);
+static uint32_t vaplic_get_ithreshold(struct vcpu *vcpu, uint16_t idc_id);
+
+void vaplic_set_hw(struct vm *vm, irqid_t intp_id)
+{
+    if (intp_id < APLIC_MAX_INTERRUPTS) {
+        bitmap_set(vm->arch.virqc.hw,intp_id);
+    }
+}
+
+/**
+ * @brief Returns if a given interrupt is associated to the physical source
+ * 
+ * @param vcpu virtual cpu running
+ * @param intp_id interrupt to evaluate
+ * @return true if is a physical intp
+ * @return false if is NOT a physical intp
+ */
+static bool vaplic_get_hw(struct vcpu* vcpu, irqid_t intp_id)
+{
+    bool ret = false;
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    if (intp_id <= APLIC_MAX_INTERRUPTS) ret = bitmap_get(virqc->hw, intp_id);
+    return ret;
+}
+
+static bool vaplic_get_pend(struct vcpu *vcpu, irqid_t intp_id){
+    uint32_t ret = 0;
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    if (intp_id < APLIC_MAX_INTERRUPTS){
+        ret = get_bit_from_reg(virqc->setip[intp_id/32], intp_id);
+    }
+    return ret;
+}
+
+static bool vaplic_get_enbl(struct vcpu *vcpu, irqid_t intp_id){
+    uint32_t ret = 0;
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    if (intp_id < APLIC_MAX_INTERRUPTS){
+        ret = get_bit_from_reg(virqc->setie[intp_id/32], intp_id);
+    }
+    return ret;
+}
+
+/**
+ * @brief Emulates the notifier aplic module.
+ *        (02/11/2022): computes the next pending bit.
+ *        Only direct mode is supported. 
+ * 
+ * @param vcpu 
+ * @return irqid_t 
+ */
+static irqid_t vaplic_emul_notifier(struct vcpu* vcpu){
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+
+    /** Find highest pending and enabled interrupt */
+    uint32_t max_prio = APLIC_MIN_PRIO;
+    irqid_t int_id = 0;
+    uint32_t hart_index = 0;
+    uint32_t prio = 0;
+
+    for (size_t i = 1; i < APLIC_MAX_INTERRUPTS; i++) {
+        if (vaplic_get_pend(vcpu, i) && vaplic_get_enbl(vcpu, i)) {
+            uint32_t target = vaplic_get_target(vcpu, i); 
+            prio = target & 0xFF; 
+            
+            if (prio < max_prio) {
+                max_prio = prio;
+                int_id = i;
+                hart_index = (target >> 18) & 0x3FFF;
+            }
+        }
+    }
+
+    /** Can interrupt be delivery? */
+    uint32_t domaincgfIE = (vaplic_get_domaincfg(vcpu) >> 8) & 0x1;
+    uint32_t threshold = vaplic_get_ithreshold(vcpu, hart_index);
+    uint32_t delivery = vaplic_get_idelivery(vcpu, hart_index);
+    uint32_t force =  vaplic_get_iforce(vcpu, hart_index);
+    if ((max_prio < threshold || threshold == 0 || force == 1) && 
```

Comment:
However, I have now noticed another problem here. Consider the following case where we arrive at this point with int_id = 1, max_prio = 1, threshold = 2, force = 1 and assuming both domain and idc are active. Although the interrupt doesn't meet the requirements to be delivered to hart, it will. This is because we do the check altogether. The simplest solution would be within this if branch to have something like this:
```
if(force){
 int_id = 0;
 max_prio = 0;
}
```
This led to another problem: in `vaplic_update_hart_line`, we are only notifying hart if int_id is != 0, as we saw int_id can be = 0 (called spurious interrupt in the specification). So I propose returning a bool value: true notifies hart, and false does not. Since we don't use the interrupt id value in `vaplic_update_hart_line` for anything else but checking whether or not it is to notify hart, it seems like the best solution to me, but I'd like to know your opinion.

Another thing that jumped out at me when reviewing this function was that it was not very readable, mainly because of the variable names, so I changed it.

In short:
```
static bool vaplic_update_topi(struct vcpu* vcpu){
    struct vaplic * vaplic = &vcpu->vm->arch.vaplic;
    bool ret = false;
    uint32_t intp_prio = APLIC_MIN_PRIO;
    irqid_t intp_id = 0;
    uint32_t intp_hart_index = 0;
    uint32_t prio = 0;
    uint32_t idc_threshold = 0;
    bool domain_enbl = false;
    bool idc_enbl = false;
    bool idc_force =  false;

    /** Find highest pending and enabled interrupt */
    for (size_t i = 1; i < APLIC_MAX_INTERRUPTS; i++) {
        if (GET_HART_INDEX(vcpu, i) == vcpu->id) {
            if (vaplic_get_pend(vcpu, i) && vaplic_get_enbl(vcpu, i)) {
                prio = vaplic_get_target(vcpu, i) & APLIC_TARGET_IPRIO_MASK; 
                if (prio < intp_prio) {
                    intp_prio = prio;
                    intp_id = i;
                    intp_hart_index = vcpu->id;
                }
            }   
        }
    }

    /** Can interrupt be delivered? */
    idc_threshold = vaplic_get_ithreshold(vcpu, intp_hart_index);
    domain_enbl = !!((vaplic_get_domaincfg(vcpu) & APLIC_DOMAINCFG_IE);
    idc_enbl = !!(vaplic_get_idelivery(vcpu, intp_hart_index));
    idc_force = !!(vaplic_get_iforce(vcpu, intp_hart_index));
    if ((intp_prio < idc_threshold || idc_threshold == 0 || idc_force) && 
         idc_enbl && domain_enbl){
        if(idc_force){
            intp_id = 0;
            intp_prio = 0;
        }
        vaplic->topi_claimi[intp_hart_index] = (intp_id << 16) | intp_prio;
        ret = true;
    } 
    return ret;
}
```

Reply:
This looks good!